### PR TITLE
[RFC 0014] Improve import from derivation

### PR DESCRIPTION
**Pull request by Ericson2314**
Fix some issues with importing from derivations so it is nicer to use both manually and with CI. If this RFC is implemented, hydra.nixos.org should allow importing from derivations.

[Rendered](https://github.com/Ericson2314/nix-rfcs/blob/import-from-derivation/rfcs/0000-import-from-derivation.md)

### CODE REVIEW

**Comment by zimbatm** [comment]
@Ericson2314 did you get my review by email? Let me know what I can do to help.

File: rfcs/0000-import-from-derivation.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+The fundamental change is rather than evaluation unconditionally triggering execution, evaluation will suspend yielding the execution needed to be done so it can make further progress on resumption.
+This is probably easier to explain in terms of the implementation steps needed.
+
+## Black-hole unwinding
+
+There's currently a bug with `nix-repl` where if computation is interrupted (say with `SIGINT`), and the same computation is begun again, one will get an infinite recursion error despite there being no infinite recursion.
+This is because the thunks stay black-holed—the indication that they are currently evaluated—despite the computation being aborted.
+To fix this, we want aborted computations to instead unwind their stack, un-black-holing any thunks back the way they were before.
+Mathematically, black-hole unwinding ensures that evaluation is idempotent.
```
Comment (zimbatm) [inline rfcs/0000-import-from-derivation.md]: Just a thought: this seems useful on it's own. Maybe split it in it's own RFC, I suspect smaller RFCs are more likely to make progress. Or just implement it?

Comment (Ericson2314) [inline rfcs/0000-import-from-derivation.md]: Seeing that that's just a bug fix, I might just open an issue. That issue can be linked from any tracking the RFC as a whole if it is accepted.

Comment (Ericson2314) [inline rfcs/0000-import-from-derivation.md]: Opened https://github.com/NixOS/nix/issues/1407

File: rfcs/0000-import-from-derivation.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+`nix-build --dry-run` is thus a synonym of `nix-build --dry-run=0`.
+`nix-build --dry-run=infinity` is the semantics of `--dry-run` today—perform as many rounds as needed to get the final static build plan.
+
+Because we pause evaluation each round, we can build the imported derivations "normally".
+That includes taking advantage of build remotes, which avoids the problem of import from derivation DOSing the hydra evaluator and or imported derivations requiring a different platform than the evaluation machine.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Overlap with recursive Nix.
```
Comment (zimbatm) [inline rfcs/0000-import-from-derivation.md]: Do you have a link to what recursive Nix is? Just so the reader can compare.

**Comment by 0xABAB** [comment]
The technical writing could be a lot better, but I get the gist of it. 

It appears that this is just an instance of "recursive make considered harmful", though. My longterm concern is that this will lead to a new workflow where people will start making use of the feature, which means that the system as a whole becomes dependent on more dynamic features while they could also be implemented with more static features (as it is now). 

In short, I would say this is a controversial feature. It's not like I can stop you, but I probably wouldn't add this feature if I had complete control over the project.

**Comment by Ericson2314** [comment]
Yes, dynamism should be the escape the escape hatch of last resort. I hope the burden of writing concrete syntax with an external program is enough to dissuade anyone from using it more.

**Comment by Ericson2314** [comment]
I should edit this RFC to also mention readFile. ReadFile of a derivation needs analogous improvements.

**Comment by domenkozar** [comment]
So if I understand correctly, to summarize: 

1) Hydra would evaluate an expression until it hits IFD and halt there
2) it would build the IFD paths
3) start with 1) using the new built paths to continue the eval

**Comment by Ericson2314** [comment]
@domenkozar do you mean " start 1) again..."? If so yes. It could also just let `nix-instantiate` do that automatically, but avoid keeping the eval process around it's probably best to do that by hand with `--dry-run`.

**Comment by edolstra** [comment]
@Ericson2314 We're doing triage on the RFC backlog, are you still interested in pursuing this RFC?

The issue I see with this RFC is that it doesn't really work given Hydra's current design, specifically, the strict split between the evaluator and the queue runner. The evaluator is supposed to add rows to the `Builds` table, so it can't delay building IFDs - it has to produce a top-level derivation. Likewise the queue runner doesn't have access to the jobset inputs, so it can't resume evaluation.

This could be solved by having an integrated evaluator/queue runner, but then we don't need any partial evaluation tricks: we can just build IFDs as normal build steps during evaluation. (The only issue is that during evaluation there is no Build yet to account the BuildStep to, but that can be hacked around by having a temporary Build or whatever.)

The downside of integrating the evaluator and queue runner is that it reduces security a bit, since (for example) the jobset Nix expression could try to read SSH credentials of the queue runner during evaluation using `readFile`.

**Comment by Ericson2314** [comment]
@edolstra Also, thank you for that description of the hydra status quo. I'll try to better separate the Nix itself and hydra concerns from this PR, but on the hydra side I'm thinking that top-level derivations must not be auto-generated, but their dependencies may be. My new idea already supported that on the Nix side, and from your description, that would be allow Hydra to keep its existing division of labor. [If we had to refactor hydra that much, I'd declare it "totalled", i.e. it would be better to just use a rewrite that would also fix a slew of other issues.]

**Comment by edolstra** [comment]
In Nix, you can't have a .drv file without knowing *all* its dependencies. This is independent from any evaluation issues.

**Comment by Ericson2314** [comment]
@edolstra yes this would be a recursive nix, thing, where drv1 would build drv2 (and drv2's deps). drv1 would be delayed evaluation of nix expressions, which we may or may not wish to special-case purely as an optimization.
